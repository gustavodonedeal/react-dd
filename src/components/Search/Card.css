:local .cardContainer {
  display: grid;
  grid-template-columns: 2fr 1fr 1fr;
  grid-template-rows: repeat(3, 1fr);
  grid-gap: 10px;
  grid-template-areas: 
    "image header header" 
    "image info info"
    "image price price";
}

:local .cardImage {
  grid-area: image;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

:local .cardHeader {
  grid-area: header;
}

:local .cardInfo {
  grid-area: info;
}

:local .price {
  grid-area: price;
}