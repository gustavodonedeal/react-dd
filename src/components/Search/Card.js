import React from 'react';
import cardStyles from './Card.css';
import { Link } from 'react-router-dom';

const CurrencySign = ({ currency }) => {
  switch (currency) {
    case 'EUR':
    default:
      return <span>â‚¬</span>;
  }
};

const Price = ({ currency, children }) => (
  <div>
    <CurrencySign currency={currency} /> {children}
  </div>
);

const Card = ({ header, age, county, price, currency, photos, id }) => (
  <Link to={`/ad/${id}`} className={cardStyles.cardContainer}>
    <img src={photos ? photos[0].large : ''} className={cardStyles.cardImage} />
    <h3 className={cardStyles.cardHeader}>{header}</h3>
    <div className={cardStyles.cardInfo}>{`${age} | ${county}`}</div>
    <h4 className={cardStyles.price}>
      <Price currency={currency}>{price}</Price>
    </h4>
  </Link>
);

export default Card;
