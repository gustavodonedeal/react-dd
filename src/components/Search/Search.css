:local .cardList {
    display: grid;
    grid-gap: 5px;
    margin: 5px;
}

:local .searchBar {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
}

:local .searchButton {
    justify-self: center;
    width: 50%;
}

:local .searchLabel {
    justify-self: center;
}