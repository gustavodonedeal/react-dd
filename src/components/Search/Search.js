import React from 'react';
import styles from './Card.css';

const SearchInput = () => <input type="text" />;

const SearchButton = () => <button>Search</button>;

const SearchBar = () =>
  <div>
    <h2>Filters</h2>
    <span>Search</span>
    <SearchInput />
    <SearchButton />
  </div>;

const SearchResult = () =>
  <div className={styles.card}>
    <img />
    <h3>Vroom</h3>
    <span>16 days | Tipp</span>
    <h4>Expensive</h4>
  </div>;

const SearchResults = ({ results }) =>
  <div>
    <h2>Cars</h2>
    <div>
      {results.map(result => <SearchResult {...result} />)}
    </div>
  </div>;

const Search = ({ results }) =>
  <div>
    <SearchBar />
    <SearchResults results={results} />
  </div>;

export default Search;
