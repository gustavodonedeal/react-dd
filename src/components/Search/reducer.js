import { getAds } from '../../api/ads';

const GET_SEARCH_RESULTS_REQUEST = 'GET_SEARCH_RESULTS_REQUEST';
const GET_SEARCH_RESULTS_SUCCESS = 'GET_SEARCH_RESULTS_SUCCESS';
const GET_SEARCH_RESULTS_FAILURE = 'GET_SEARCH_RESULTS_FAILURE';
const SET_SEARCH_FILTER = 'SET_SEARCH_FILTER';

export const requestSearchResults = () => ({
  type: GET_SEARCH_RESULTS_REQUEST
});
export const searchResultsSuccess = payload => ({
  type: GET_SEARCH_RESULTS_SUCCESS,
  payload
});
export const searchResultsFailure = error => ({
  type: GET_SEARCH_RESULTS_FAILURE,
  error
});

export const setSearchFilter = filter => ({
  type: SET_SEARCH_FILTER,
  filter
});

export const fetchSearchResults = (filter = '') => dispatch =>
  new Promise(async (resolve, reject) => {
    dispatch(requestSearchResults());
    try {
      const adverts = await getAds(filter);
      resolve(dispatch(searchResultsSuccess(adverts)));
    } catch (error) {
      console.log(error);
      reject(dispatch(searchResultsFailure(error)));
    }
  });

export const initialState = {
  loading: false,
  filter: '',
  results: [],
  error: null
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case GET_SEARCH_RESULTS_REQUEST:
      return Object.assign({}, state, {
        loading: true,
        error: null,
        results: []
      });
    case GET_SEARCH_RESULTS_SUCCESS:
      return Object.assign({}, state, {
        results: action.payload,
        loading: false,
        error: null
      });
    case GET_SEARCH_RESULTS_FAILURE:
      return Object.assign({}, state, {
        loading: false,
        error: action.error
      });
    case SET_SEARCH_FILTER:
      return Object.assign({}, state, {
        filter: action.filter
      });
    default:
      return state;
  }
}

export const selectors = {
  isLoading: state => state.loading,
  getSearchFilter: state => state.filter,
  getSearchAds: state => state.results.ads || []
};
