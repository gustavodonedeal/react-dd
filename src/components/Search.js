import React, { Component } from 'react';
import axios from 'axios';
import '../App.css';
import Card from './Card';
import Filters from './Filters';

class Search extends Component {

  constructor() {
    super();
    this.loadAds = this.loadAds.bind(this);
    this.updateAds = this.updateAds.bind(this);

    // start the state
    this.error = null;
    this.section = 'cars';

    axios.defaults.headers.post['Content-Type'] = 'application/json';
  }

  componentDidMount() {
    setTimeout(() => {
      this.section = 'cars';
      //this.section = this.props.match.params.section || 'cars';
      if (this.props.loading && !this.props.ads.length) {
        this.loadAds();
      }
    }, 0);
  }

  loadAds() {
    const cors = 'https://cors-anywhere.herokuapp.com/';
    const params = Object.assign({}, {
      adType: 'forsale',
      max: 30,
      section: this.section,
      sort: 'relevance desc',
    }, this.props.filter) ;

    this.serverRequest = axios.post(`${cors}https://www.donedeal.ie/search/api/v4/find/`, params)
      .then((response) => {
        this.props.updateState({
          ads: response.data.ads,
          loading: false
        });
      }).catch((error) => {
        console.error(error);
      });
  }

  updateAds(filter) {
    this.props.updateState({
      ads: [],
      loading: true,
      filter: filter
    });
  }

  render() {
    const { ads, loading } = this.props;
    let cards;

    if(loading) {
      return (<p>Loading...</p>)
    } else if (!ads) {
      cards = (
        <p>{`No results for your search ðŸ˜ž`}</p>)
    } else {
      cards = (
        <ul>
          {ads.map((ad) =>
            <Card ad={ad} key={ad.id} />
          )}
        </ul>
      )
    }

    return (
      <div className="App">
        <Filters updateAds={this.updateAds} />
        <h1>{this.section.toUpperCase()}</h1>
        {cards}
      </div>
    );
  }
}

export default Search;
