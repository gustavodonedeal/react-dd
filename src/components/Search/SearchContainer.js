import React, { Component } from 'react';
import { connect } from 'react-redux';
import Search from './Search';
import { selectors, setSearchFilter, fetchSearchResults } from './reducer';

class SearchContainer extends Component {
  async componentDidMount() {
    const { results, getSearchResults, filter } = this.props;
    if (results.length === 0) {
      const data = await getSearchResults(filter);
    }
  }

  async componentDidUpdate(prevProps) {
    const { filter, getSearchResults } = this.props;
    if (filter !== prevProps.filter) {
      const data = await getSearchResults(filter);
      console.log(data);
    }
  }

  render() {
    return <Search {...this.props} />;
  }
}

const mapStateToProps = state => ({
  loading: selectors.isLoading(state.search),
  results: selectors.getSearchAds(state.search),
  filter: selectors.getSearchFilter(state.search)
});

const mapDispatchToProps = dispatch => ({
  getSearchResults: () => dispatch(fetchSearchResults()),
  onFilter: event => dispatch(setSearchFilter(event.target.value))
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchContainer);
