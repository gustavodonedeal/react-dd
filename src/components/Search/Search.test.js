import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import fetchMock from 'fetch-mock';
import Search from './Search';

const mockResponse = (status, statusText, response) => {
  return new window.Response(response, {
    status: status,
    statusText: statusText,
    headers: {
      'Content-type': 'application/json'
    }
  });
};

it('renders "No Results found" correctly', async () => {
  const json = await import('../../../__tests__/sampleAdvert.json');
  fetchMock.mock('*', [json]);
  const tree = renderer.create(
    <MemoryRouter>
      <Search />
    </MemoryRouter>
  ).toJSON();
  expect(tree).toMatchSnapshot();
  fetchMock.restore();
});