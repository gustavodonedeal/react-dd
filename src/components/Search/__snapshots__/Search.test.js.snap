// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders "No Results found" correctly 1`] = `
<p>
  Loading...
</p>
`;
